lightning/src/ln/channelmanager.rs

```diff
@@ -5301,11 +5420,33 @@ impl<'a, Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 			}
 		}
 
-		let pending_outbound_payments_count: u64 = Readable::read(reader)?;
-		let mut pending_outbound_payments: HashSet<[u8; 32]> = HashSet::with_capacity(cmp::min(pending_outbound_payments_count as usize, MAX_ALLOC_SIZE/32));
-		for _ in 0..pending_outbound_payments_count {
-			if !pending_outbound_payments.insert(Readable::read(reader)?) {
-				return Err(DecodeError::InvalidValue);
+		let pending_outbound_payments_count_compat: u64 = Readable::read(reader)?;
+		let mut pending_outbound_payments: HashMap<MppId, HashSet<[u8; 32]>> =
+			HashMap::with_capacity(cmp::min(pending_outbound_payments_count_compat as usize, MAX_ALLOC_SIZE/32));
+
+		match pending_outbound_payments_count_compat {
+			0 => {
```

Comment:
Old clients can have zero entries in their pending outbound payment sets, in fact are likely to. Instead, I think we should read the old entries, insert them as needed, and read the new entries via a TLV.

Reply:
That'd mean adapting the TLV macros to work with hashmaps, right?

Reply:
That would just mean implementing `Readable`/`Writeable` for hashmaps, I don't think any changes to the macros needs to happen. Just write a backwards-compat read block, then try to read an `optional` HashMap in the TLV read and if it comes back `Some`, overwrite the backwards-compat read.